File: packages/api-contracts/src/report/render.ts
import { z } from "zod";
import { LocaleSchema } from "../locale.js";
import { ModuleIdSchema, RoleIdSchema, SafetyLevelIdSchema } from "../morpheo/enums.js";

/**
 * The report service boundary (build plan §5.6 / §20 Checkpoint 11.1). The
 * report renders an ALREADY-COMPUTED Morpheo result into a versioned, localized
 * document; it never recalculates scores, alters safety flags, or invents
 * clinical facts. The clinical wording is Morpheo's approved content — the
 * report only lays it out.
 */

/**
 * edge/worker -> report: render request. Carries the deterministic result the
 * report lays out (the build plan's safetyFlags + orientationCodes), plus the
 * versions stamped on the output and the locale to render in.
 */
export const ReportRenderRequestSchema = z
  .object({
    assessmentId: z.string().min(1),
    // Morpheo workflow_version (the rule/spec version) and content_version.
    definitionVersion: z.string().min(1),
    contentVersion: z.string().min(1),
    locale: LocaleSchema,
    role: RoleIdSchema,
    // safetyFlags: the fired safety outcome (never recomputed here).
    level: SafetyLevelIdSchema.nullable(),
    stop: z.boolean(),
    triggeredRules: z.array(z.string().min(1)),
    // orientationCodes: the activated clinical modules.
    routes: z.array(ModuleIdSchema),
    completedAt: z.string().min(1),
  })
  .strict();
export type ReportRenderRequest = z.infer<typeof ReportRenderRequestSchema>;

/**
 * report -> edge: an immutable reference to a rendered report. The signed URLs
 * are short-lived and served through the edge (§9); they are null until the
 * render+store step exists.
 */
export const ReportRefSchema = z
  .object({
    reportId: z.string().min(1),
    assessmentId: z.string().min(1),
    templateVersion: z.string().min(1),
    definitionVersion: z.string().min(1),
    contentVersion: z.string().min(1),
    locale: LocaleSchema,
    createdAt: z.string().min(1),
    htmlUrl: z.string().min(1).nullable(),
    pdfUrl: z.string().min(1).nullable(),
  })
  .strict();
export type ReportRef = z.infer<typeof ReportRefSchema>;

/** The named set exported as JSON Schema artifacts (schemas/json-schema/report). */
export const REPORT_CONTRACT_SCHEMAS = {
  ReportRenderRequest: ReportRenderRequestSchema,
  ReportRef: ReportRefSchema,
} as const;

File: services/somnus-report-service/src/report/schemas/render.py
"""Report render DTOs (build plan §20 Checkpoint 11.1).

`ReportRenderRequestDTO` mirrors the edge↔report Zod contract (validated against
the shared JSON Schema by the contract test). `ClinicalContentDTO` is the
approved Morpheo wording the report lays out — the report never authors it.
"""

from __future__ import annotations

from typing import Literal

from report.schemas.base import ContractModel

Role = Literal["adult", "parent", "professional"]
SafetyLevel = Literal["L0", "L1", "L2", "L3", "L4"]
Module = Literal["INS", "BRE", "SLP", "CIR", "RLS", "PAR"]
Locale = Literal["es", "en", "ca", "fr"]


class ReportRenderRequestDTO(ContractModel):
    assessment_id: str
    definition_version: str
    content_version: str
    locale: Locale
    role: Role
    level: SafetyLevel | None
    stop: bool
    triggered_rules: list[str]
    routes: list[Module]
    completed_at: str


# --- The approved Morpheo content the report lays out (subset of the content
# endpoint's response; the report only presents it, never restates it). ---


class SafetyLevelContentDTO(ContractModel):
    id: SafetyLevel
    name: str
    action: str


class ModuleContentDTO(ContractModel):
    id: Module
    name: str
    minimum_questions: list[str]
    output: str


class OutputContractContentDTO(ContractModel):
    patient_parent: list[str]
    professional: list[str]
    forbidden_phrases: list[str]


class ClinicalContentDTO(ContractModel):
    locale: Locale
    content_version: str
    modules: list[ModuleContentDTO]
    safety_levels: list[SafetyLevelContentDTO]
    output_contract: OutputContractContentDTO


class ReportRefDTO(ContractModel):
    report_id: str
    assessment_id: str
    template_version: str
    definition_version: str
    content_version: str
    locale: Locale
    created_at: str
    html_url: str | None
    pdf_url: str | None
